### PR TITLE
added resnet model

### PR DESCRIPTION
**Pull request by dnabanita7**
I have made changes as per review.

### CODE REVIEW

File: src/VisionModel/resnet.jl [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    push!(layers, Conv((7, 7), 3=>inplanes, stride=(2, 2), pad=(3, 3)))
+    push!(layers, BatchNorm(inplanes, λ=relu))
+    push!(layers, MaxPool((3, 3), stride=(2, 2), pad=(1, 1)))
+    inplanes = 64
+    outplanes = 64
+    for nrepeats in block_config
+    
+      push!(layers, SkipConnection(basicblock(inplanes, outplanes, downsample), 
+                          projection(inplanes, planes, stride)))
+      outplanes = outplanes * 2
```
Comment (darsnack) [inline src/VisionModel/resnet.jl:48]: We won't need this block. All the logic will be in the inner most loop.

File: src/VisionModel/resnet.jl [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+      for expansion in channel_config
+        push!(layers, SkipConnection(bottleneck(inplanes, outplanes, downsample),
```
Comment (darsnack) [inline src/VisionModel/resnet.jl:50]: Before this line, `outplanes = inplanes * expansion`.

File: src/VisionModel/resnet.jl [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    expansion = 1
+    stride = 1
+    end =#
+    layers = []
+    push!(layers, Conv((7, 7), 3=>inplanes, stride=(2, 2), pad=(3, 3)))
+    push!(layers, BatchNorm(inplanes, λ=relu))
+    push!(layers, MaxPool((3, 3), stride=(2, 2), pad=(1, 1)))
+    inplanes = 64
+    outplanes = 64
+    for nrepeats in block_config
```
Comment (darsnack) [inline src/VisionModel/resnet.jl:44]: We actually need to repeat the block `nrepeats` times. So the overall structure will look like:
```julia
for nrepeats in block_config
  for i in 1:nrepeats
    for expansion in channel_config
      # add a SkipConnection
    end
  end
  inplanes *= 2
end
```

File: src/VisionModel/resnet.jl (hunk 2 above)
Comment (darsnack) [inline src/VisionModel/resnet.jl:50]: Here, for `downsample`, pass `i == 1`. We want to downsample when we start a new repetitions of blocks.

File: src/VisionModel/resnet.jl [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    
+      push!(layers, SkipConnection(basicblock(inplanes, outplanes, downsample), 
+                          projection(inplanes, planes, stride)))
+      outplanes = outplanes * 2
+      for expansion in channel_config
+        push!(layers, SkipConnection(bottleneck(inplanes, outplanes, downsample),
+                                     projection(inplanes, outplanes, stride)))
+        inplanes = inplanes * expansion
```
Comment (darsnack) [inline src/VisionModel/resnet.jl:52]: You want `inplanes` to stay at whatever value comes into a repetition of blocks, so you can remove this line.

File: src/VisionModel/resnet.jl [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    inplanes = 64
+    outplanes = 64
+    for nrepeats in block_config
+    
+      push!(layers, SkipConnection(basicblock(inplanes, outplanes, downsample), 
+                          projection(inplanes, planes, stride)))
+      outplanes = outplanes * 2
+      for expansion in channel_config
+        push!(layers, SkipConnection(bottleneck(inplanes, outplanes, downsample),
+                                     projection(inplanes, outplanes, stride)))
```
Comment (darsnack) [inline src/VisionModel/resnet.jl:51]: This should be the identity shortcut for ResNet-50 (not projection).

**Review by darsnack — CHANGES_REQUESTED** [review]
Okay I think this is good. The only other change after this is that we designed the loop to work for ResNet-50+. So it always uses bottleneck blocks and identity shortcuts. We will want to modify `resnet` so that it can use any combination of blocks + shortcuts. This will allow the ResNet-18 and ResNet-34 configurations to work as well.

File: src/VisionModel/resnet.jl (hunk 5 above)
Comment (dnabanita7) [reply src/VisionModel/resnet.jl:51]: there is no ``projection function`` in resnet. All we have is and ``identity`` shortcut.

**Comment by dnabanita7** [comment]
I have deleted the underlying MLH branch as Rohan said it is fine to not use the MLH forks so I will be directly pushing from my fork further on. I know its annoying and I apologize for that but I am closing this here and opening a new one.

**Pull request closed without merging**